velox/docs/develop/memory.rst

```diff
@@ -0,0 +1,890 @@
+the largest capacity in the system. The used memory reclaim is achieved through
+techniques like `disk spilling <https://facebookincubator.github.io/velox/develop/spilling.html>`_. If the memory arbitration succeeds, the leaf
+operator pool can proceed with the actual memory allocation from memory
+allocator. However, if the memory arbitration fails, then the query which has
+the largest capacity in the system is chosen to fail with query memory capacity
+exceeded error (local OOM). The failed query may or may not be the requestor
+query pool itself.
+
+The memory allocator does the actual memory allocation from its own managed
+memory space in units of machine pages (4KB). It tracks the amount of allocated
+memory, and returns an error if an allocation request exceeds the system memory
+limit. This enables explicit control on RSS of Velox memory usage to help
+prevent the server OOM.
+
+The file cache provides the in-memory hot data cache and prefetch functions
+when the user query accesses the remote storage. It allocates memory from the
+memory allocator directly which is not counted in query memory usage. To
+prevent the memory allocation failure because of excessive file cache memory
+usage, the file cache retries the allocation failure by shrinking the file
+cache. This achieves dynamic memory sharing between file cache and query memory
+in response to the user query workload changes.
+
+.. image:: images/memory-function.png
+    :width: 800
+    :align: center
+    :alt: Memory Management Functions
+
+To summarize, the memory manager manages the memory pools and coordinates the
+accesses between different memory components. The memory pool tracks a query’s
+memory usage and interacts with the memory arbitrator to adjust the memory
+capacity allocations among running queries to achieve fair memory sharing. The
+memory allocator manages the physical memory allocations to prevent server OOM,
+and interacts with file cache to achieve dynamic memory sharing between query
+memory and file cache to maximize the memory efficiency. The rest of the
+document describes each memory component in detail.
+
+Memory Manager
+--------------
+
+.. image:: images/memory-manager.png
+    :width: 600
+    :align: center
+    :alt: Memory Manager
+
+The memory manager is created on server startup with provided
+*MemoryManagerOption*. It creates a memory allocator instance to manage the
+physical memory allocations for both query memory allocated through memory pool
+and cache memory allocated through the file cache. It ensures the total
+allocated memory is within the system memory limit (specified by
+MemoryManagerOptions::allocatorCapacity*). The memory manager also creates a
+memory arbitrator instance to arbitrate the memory capacity among running
+queries. It ensures the total allocated query memory capacity is within the
+query memory limit (specified by *MemoryManagerOptions::arbitratorCapacity*). The
```

Comment:
Maybe include a digram showing relationship between different limits: allocatorCapacity vs. arbitratorCapacity vs. total memory available for the whole process.

Reply:
Good point, I will add one in followup.